### PR TITLE
Add Chinese error message translations

### PR DESCRIPTION
**Pull request by MichaelChirico**
Something I've been meaning to figure out for years & finally figured out. The hard part (setting up the `po` directory) is done -- the rest is the manual work of actually translating strings.

Before proceeding, I wanted to file this and get some feedback about whether it's worth to proceed.

I see benefits:

- Extremely welcoming to the Chinese-speaking R community which is quite large and counts among its members several of our biggest contributors & advocates. Of course they (i.e. the contributors) all speak English just fine & understand the current messaging, but they may be (hopefully I'm not presuming too much here) more productive if they can get native feedback without the extra level of translation.
- As near as I can tell, we would be the _only_ of the [top 100 or so](https://www.r-pkg.org/starred) R packages to offer translation to _any_ language (none of them have the required `po` directory). This would make `data.table` a leader in accessibility and openness in yet another dimension for the R community.
 
Costs:

- Maintenance burden -- _should_ be updating the main files as new errors/messages are added in `R/` and `src/` going forward. Hopefully minimal, but it's new territory -- we'd be one of the only major packages stress testing the tools R provides here. See steps below for a bit more detail.
- Package size -- We're already lapping at the 5MB CRAN limit basically every time we submit recently. I don't _think_ the `po/` directory is included in the built package but it would be 1000s of lines if so. The compressed file added to `inst/` is definitely included but is only 500B as of now... for comparison, the entirety of `r-source` has about 6.3Mb of `.mo` files (`sum(file.info(list.files('~/github/r-source', pattern = '.mo$', recursive = TRUE, full.names = TRUE))$size)/1e6`)... but I think 5Kb is a very conservative estimate of how big our file could get even with several other translations included.
- Potential trouble with new issues filed with unknown messages -- solution is to provide instructions in the Issues template for reporters on how to convert their messages back to English, but not clear how effective this would be

Any other missed considerations please file below and/or edit into this post.

cc @renkun-ken @shrektan @dracodoc -- if you're willing to put any time / effort into filling out some translations please signal so here as well 😄 If we move forward we can also post on Twitter where I know we have several more Chinese/Chinese diaspora followers.

The last consideration is -- why Chinese?

It's really about perceived utility & popularity. We can of course post a poll on Twitter before proceeding (or more objectively, does CRAN offer statistics on IPs downloading?); the other two choices would seem to be Spanish or Japanese.

---

For my own reference as much as anything, here are some notes/steps to getting the package to a portable state (from a package with no portability, i.e. no `po` directory). [WRE has a few sections](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#C_002dlevel-messages) but I read them many times before figuring this all out (`data.table` can always be replaced by a generic other package below):

1. In root directory, run `tools::update_pkg_po('.')`. This will create the initial `po` folder and the `po/R-data.table.pot` file which has extracts of all arguments to `stop`/`error`/`message` used in the `R` directory.
2. Portability for messages in `src` is a lot more manual -- any message that's to be translated needs to be wrapped to the macro function `_` (e.g. `error("this error")` is edited to `error(_("this error"))`), then we run `xgettext --keyword=_ -o data.table.pot *.c`, which will create a file analogous to `R-data.table.pot` for C-level messages. Some useful command line stuff for this:

```
# I ran this for loop to append `_`
for SRC_FILE in src/*.c;
  # no inplace -i in default mac sed
  do sed -E 's/error[(]("[^"]*")/error(_(\1)/g' $SRC_FILE > out;
  mv out $SRC_FILE;
done
# checking for other lines calling error() that didn't get _()-wrapped
grep -Er '\berror[(]' src --include=*.c | grep -v _ | grep -Ev '(?://|[*]).*error[(]'

# ditto for warning, DTWARN, DTPRINT, Rprintf, STOP, Error, mutatis mutandis
#   a bit more manual was snprintf usage
# NB that this approach didn't work perfectly, there are a few edge cases that I caught when
#   installation failed & had to fix manually:
#   * quote embedded (and escaped) within message [could be fixed with smarter regex]
#   * multi-line implicit-concat arrays (in C, `"a" "b"` is the same as `"ab"`) should be wrapped "on the outside" not individually
#   * `data.table` shares some of its `src` with `pydatatable`, so the requirement to `#include <R.h>` before the `#define _` macro meant we need to be careful about including this macro only in the R headers for these files (hence I created `po.h`)
#   * Can't use `_()` _inside_ another functional macro. Only wrap the string passed to the macro later.

# look for char array that haven't been covered yet
grep -Er '"[^"]+"' src --include=*.c | grep -Fv '_("' | grep -v "#include" | grep -v '//.*".*"'

# look for lines starting with a char array (likely continued from prev line & can be combined)
grep -Er '^\s*"' src/*.c
```

3. Create a `.po` file in `po/` corresponding to a target language, e.g. `touch po/R-zh_CN.po` for Mandarin. `R-` means it corresponds to the R messages; see the `po/` directories in e.g. `r-source/src/library/{base,stats,utils,tools}` for locales already in use by `base`, or see [the R manual](https://cran.r-project.org/doc/manuals/r-release/R-admin.html#Localization-of-messages) for a more official documentation of these codes.
4. Add a header to the file like so (need to edit `PO-Revision-Date`, `Last-Translator`, `Language-Team`, `Language` and the blank `msgid` & `msgstr` lines are necessary):

```
msgid ""
msgstr ""
"Project-Id-Version: data.table 1.12.5\n"
"POT-Creation-Date: 2019-10-04 17:06\n"
"PO-Revision-Date: 2019-10-04 17:06+08\n"
"Last-Translator: Michael Chirico <MichaelChirico4@gmail.com>\n"
"Language-Team: Mandarin\n"
"Language: Mandarin\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
```

5. Run `tools::update_pkg_po('.')` again; this time, the captured messages and blank translations will be added to your `.po` file.

6. Add translations where the blank strings are in the `.po`. Each string corresponds to an argument of e.g. `stop`, so e.g. `stop("hi", "you", "broke", "it")` in `R/` would create 4 entries in the `.pot`&`.po` files. This may make translation more difficult if the target language has sufficiently different grammar that the components can't be made to line up in order & make sense. I think the solution here is to use `gettext`/`ngettext` but I'm not sure.

7. Save & run `tools::update_pkg_po('.')` again. This time, the `.po` file will be processed and the translations will be recorded in the `.mo` file in `inst/po`.

8. Reinstall `data.table` and run `R` in the correct domain. This can be done _ad hoc_ by temporarily setting the `LANGUAGE` env, e.g. `LANGUAGE=zh_CN.UTF-8 R --no-save` will start R in Chinese without editing the locale on your machine more deeply. Then trigger a few of the messages to make sure the process was successful.

*NB* - there was a [bug in `tools`](https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17606) leading `update_pkg_po` to fail that was finished (as of this writing) extremely recently. If you encounter this bug, either install bleeding-edge `R-devel` or simply assign the corrected version of `tools:::en_quote` and `tools::update_pkg_po` to your `.GlobalEnv` (as well as a few more functions internal to `tools`) & run from console instead of using the `tools` version.

---

I think it will also be helpful to add a note to the start-up message about where how to find help _from_ non-English _to_ English. One way is to provide a link to the `.po` files about how to convert the error message to English before proceeding. Or perhaps numbering errors? I'm not sure the best way here.

### CODE REVIEW

**Comment by dracodoc** [comment]
I would love to contribute on my favorite and most used package!
I will comment more when I got time to look at the details.

**Comment by shrektan** [comment]
No problem I am at your command

**Comment by renkun-ken** [comment]
Thanks for your research on this. I'd love to contribute!

**Comment by dracodoc** [comment]
About the package size, will it be possible to separate language files into separate download?
- Many software have this design, by default only load 1-2 languages, then use can download language files.
- If we host the language file outside of CRAN, all the languages will not cause add any burden. It's also more friendly to users that they don't need to download unneeded languages.
- After the default package installation, there could be a multiple language prompt that tell user how to use a helper function to download language files.

**Comment by dracodoc** [comment]
I haven't get time to look at how this is implemented yet, though if the translation task do need multiple human hours, it will be ideal that the translation task can be distributed to volunteer translators that with minimal setup requirement (I'm expecting translators should at least quite familiar with R and data.table, but the steps mentioned above can be a little bit overwhelming).

Is it possible to create a table and translators can just fill in the table? One organizer can distribute the tasks to translators, standardize the frequently used words/concepts. This will make the translation part much easier. 

Then there could be some script to read the table and create the needed files.

I'm not sure if this is possible or easy to implement, since I haven't look at the details.

**Comment by MichaelChirico** [comment]
Yep, that's basically what I have in mind! First I want to reach out on Twitter for more potential translators, just waiting to hear some first reactions from @jangorecki & @mattdowle

**Comment by mattdowle** [comment]
Sounds good to me! In #3937 I reduced installed size down by 1MB to about 4.0MB.  You mentioned a few size figures for `po`, one being just 500 bytes so even at 100x that should be fine. If we had multiple languages, a user typically (by default) doesn't install all translations I assume, so even if CRAN size checks do count total size of all languages in `po` it's a problem we can raise with them later if and when we reach it.

**Comment by jangorecki** [comment]
Idea is good.
English should generally be preferred, it is much easier to google for error messages, or sometimes even understand the message in English. I recall it was annoying when I had to configure R to english rather than polish.
I think the idea is good because in some countries English has quite low adoption, even among technical users. Chinese, Russian, Spanish would be definitely useful here.

**Comment by MichaelChirico** [comment]
> it is much easier to google for error messages

Yes, this is a sticking point. I'm not sure the best way around this. Right now, I think a good approach is to add a new start-up message that can point non-native users to the `.po` files to look up the English message for Googling & filing issues?

Open to other suggestions here.

**Comment by dracodoc** [comment]
For Google search error message, is it practical to assign an error number to every message? 
Sometimes same error can have different wordings among versions which also make search harder. Having a fixed error number will also help on this case.

**Comment by MichaelChirico** [comment]
OK here are some of my thoughts for how to proceed. 

I prefer to keep pretty focused for now. As mentioned we're the first major R package to try this so we're in unexplored territory -- I don't want to venture too far outside the status quo for package dev while we figure out what are the tradeoffs of adding non-English support. So, as much as I think things like a website, wiki, etc. added in Chinese are quite useful, let's pause on that for now. 

Re: adding numbered messages, I lean towards not doing it. I'm not sure how to scale that or how practical it is to do at this stage. (1) 15+ years of `data.table` help is available online with no numbering (2) adding numbering to every error message would make the diff of this PR much harder to sift through (3) should we be numbering R & C errors separately? (4) Suppose we came up with a numbering system for error messages today. How do we number the next error that gets added? Will we have to keep track of the current error number somewhere? What about when multiple PRs are adding different errors, does the maintainer (i.e., Matt) have to de-conflict the error numbers and guard against duplicates there too? (I have in mind this similar endeavour where duplications crept in to our manual test numbering system over the years: https://github.com/Rdatatable/data.table/pull/3539 And that is all in a single file! Much harder to do over many scripts and between R & C code). So, I think the more practical thing to do is to add to the start-up message some instructions in native language on how to find the English version of errors/messages.

Lastly, I think that coordinating everything here on GitHub will get inconvenient/impractical very quickly. Instead, I propose to use a (through-now dormant) Slack channel to simplify things. This is for convenience/dev speed, however, I want to keep as transparent as possible with this (1) because of the importance of transparency in FOSS in general and (2) because such archive can serve quite useful to other R teams/packages considering going a similar route in the future. Therefore, all chats should happen in public channels, and at the conclusion (or maybe regularly -- if the volume of messages is large, we'll have to download occasionally given the free messaging limits in Slack), I'll post the full archive here as a textfile.

Current team of volunteers is: @renkun-ken @shrektan @dracodoc @hongyuanjia @Leo-Lee15 @amy17519 @zhiiiyang @GuangchangYu

Please contact me at my gmail address: MichaelChirico4 and I'll invite you to the channel.

Thanks so much everyone! It's been very encouraging to see the outpouring of support :)

**Comment by jangorecki** [comment]
Where is `_()` C function defined? would be useful to separate big changes into different commits, so changes to all `error(` calls separately from `_()` definition, even separately from `pot` files. Then it is easy to filter out changes to review. Do you have a minimal working example somewhere? Trying to make it work on DT from start might be much more complex than first having a minimal example. One error from R, one error from C, and translations for them.

**Comment by MichaelChirico** [comment]
> Where is _() C function defined?

This latest commit was simply applying section 1.8.1 of WRE:

https://cran.r-project.org/doc/manuals/r-release/R-exts.html#C_002dlevel-messages

I've added this step to the instructions above. I think this makes sense as a single commit, though it does still need touch-up

**Comment by jangorecki** [comment]
@MichaelChirico I don't think we should translate verbose messages, errors and warnings should be enough.

File: src/dogroups.c [hunk 1]
```diff
@@ -11,13 +11,13 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
   Rboolean wasvector, firstalloc=FALSE, NullWarnDone=FALSE;
   clock_t tstart=0, tblock[10]={0}; int nblock[10]={0};
 
-  if (!isInteger(order)) error("Internal error: order not integer vector"); // # nocov
-  if (TYPEOF(starts) != INTSXP) error("Internal error: starts not integer"); // # nocov
-  if (TYPEOF(lens) != INTSXP) error("Internal error: lens not integer"); // # nocov
-  // starts can now be NA (<0): if (INTEGER(starts)[0]<0 || INTEGER(lens)[0]<0) error("starts[1]<0 or lens[1]<0");
-  if (!isNull(jiscols) && LENGTH(order) && !LOGICAL(on)[0]) error("Internal error: jiscols not NULL but o__ has length"); // # nocov
-  if (!isNull(xjiscols) && LENGTH(order) && !LOGICAL(on)[0]) error("Internal error: xjiscols not NULL but o__ has length"); // # nocov
-  if(!isEnvironment(env)) error("’env’ should be an environment");
```
Comment (MichaelChirico) [inline src/dogroups.c]: angled quotes are non-ASCII, failed `xgettext`

**Comment by MichaelChirico** [comment]
Any reason why not? Presumably non-English would benefit from getting their verbose output in native as well. Marginal burden of translation seems manageable.

With these latest commits I was more concerned about the use of `_` in files that are used in pydatatable. @st-pasha does portability work similarly in Python, or should we have an identity `#define` of `_` for Python? Or is that something you can manage in your headers?

**Comment by jangorecki** [comment]
Verbose messages as not the output that regular users see. It increase the amount of work on the translations and maintenance. At least at start I don't think we should start with translating verbose messages.

**Comment by MichaelChirico** [comment]
That's fair, but it's a bit tough practically to separate verbose from non-verbose messages

**Comment by MichaelChirico** [comment]
Just added a few commits increasing coverage on `fmelt.c` since this was failing CI (edit error messages for translation -> looks like missed diff in codecov -> fails).

**Comment by MichaelChirico** [comment]
Attaching here the transcript of the working group Slack channel, Oct 12-30, 2019. 

[zh-translation-20191012-30.txt](https://github.com/Rdatatable/data.table/files/3793253/zh-translation-20191012-30.txt)

File: R/onAttach.R [hunk 2]
```diff
@@ -20,6 +20,8 @@
   if (!isTRUE(getOption("datatable.quiet"))) {   # new option in v1.12.4, #3489
     packageStartupMessage("data.table ", v, if(dev)paste0(" IN DEVELOPMENT built ",d,g),
                           " using ", getDTthreads(verbose=FALSE), " threads (see ?getDTthreads).  Latest news: r-datatable.com")
+    if (gettext("TRANSLATION CHECK", domain='R-data.table') != "TRANSLATION CHECK")
+      packageStartupMessage(gettext("**********\n", "Running data.table in English; package support is available in English only. When searching for online help, be sure to also check for the English error message. This can be obtained by looking at the po/R-<locale>.po and po/<locale>.po files in the package source, where the native language and English error messages can be found side-by-side"))
```
Comment (mattdowle) [inline R/onAttach.R]: Should that English be non-English?
If so, how about more detail:
```data.table is running in non-English mode; i.e. errors and message will be displayed in Chinese.  When reporting problems or searching for help though, please use the English translation which can be found in data.table's source `po` directory:  po/R-<locale>.po and po/<locale>.po. You can also start data.table/R in English mode by doing ...```

Comment (MichaelChirico) [reply R/onAttach.R]: if this error is triggered, it will be translated; in Chinese it'll say translate. In future, it could say other languages.

**Comment by MichaelChirico** [comment]
It's still WIP. Have left 2 more weeks to translate & 2 more weeks after that for reviews; progress can be checked here:

https://docs.qq.com/sheet/DVkluUWRLb0ttYkFL?tab=BB08J2&c=B4A0A0

(ICYMI I've been merging the other PRs against this branch, I had in mind to merge only this branch at the end into master)

**Comment by mattdowle** [comment]
That spreadsheet looks like a big undertaking and well organized!
Do iiuc that all the yellow translation PRs hang off this one. You're going to merge them into this one? Only reason I'm asking is just to check I shouldnt be looking at or merging anything tagged translation.
Is there any piece of translation you want to get into this release in the next day or so? If it's possible to release a small first step to CRAN now then it can often help to reveal unexpected errors that can be fixed in the inevitable follow up. Since it takes a week or two for all CRAN instances to run. I can imagine the nature of it is all-or-nothing though.

**Comment by MichaelChirico** [comment]
> Do iiuc that all the yellow translation PRs hang off this one. You're going to merge them into this one?

Yep that's right.

> I can imagine the nature of it is all-or-nothing though.

We _could_ merge the progress so far (about 50%), I think it's best to save it so all the initial translation stuff is part of one release.

File: src/fread.c [hunk 3]
```diff
@@ -1481,15 +1485,15 @@ int freadMain(freadMainArgs _args) {
...
-              DTPRINT("  with %d fields using quote rule %d\n", topNumFields, quoteRule);
+              DTPRINT(_("  with %d fields using quote rule %d\n"), topNumFields, quoteRule);
             }
           }
         } else {
           // standard regular same-ncol-per-row file, possibly with a meta header to be skipped
           const char *prevLineStart=ch, *lineStart=ch;
           int lastncol = countfields(&ch);
           if (lastncol<0) continue;  //   invalid file with this sep and quote rule, skip
-          ASSERT(lastncol>0, "Internal error: first non-empty row should always be at least one field; %c %d", sep, quoteRule); // # nocov
```
Comment (MichaelChirico) [inline src/fread.c]: note: have removed `Internal error` here since `ASSERT` macro already has that

File: src/init.c [hunk 4]
```diff
@@ -206,59 +206,59 @@ void attribute_visible R_init_datatable(DllInfo *info)
...
-  if ((uint_fast8_t)('9'-':') < 10) error("The C expression (uint_fast8_t)('9'-':')<10 is true. Should be false.");
+  if ((uint_fast8_t)('0'-'/') != 1) error(_("The ascii character '/' is not just before '0'"));
+  if ((uint_fast8_t)('/'-'0') < 10) error(_("The C expression (uint_fast8_t)('/'-'0')<10 is true. Should be false."));
+  if ((uint_fast8_t)(':'-'9') != 1) error(_("The ascii character ':' is not just after '9'"));
+  if ((uint_fast8_t)('9'-':') < 10) error(_("The C expression (uint_fast8_t)('9'-':')<10 is true. Should be false."));
 
   // Variables rather than #define for NA_INT64 to ensure correct usage; i.e. not casted
   NA_INT64_LL = LLONG_MIN;
   NA_INT64_D = LLtoD(NA_INT64_LL);
-  if (NA_INT64_LL != DtoLL(NA_INT64_D)) error("Conversion of NA_INT64 via double failed %"PRId64"!=%"PRId64"", (int64_t)NA_INT64_LL, (int64_t)DtoLL(NA_INT64_D));
```
Comment (MichaelChirico) [inline src/init.c]: removed the redundant `""` here

**Comment by jangorecki** [comment]
Vistors to r-datatable.com from China are 6% of traffic last month, second largest user base after US (27%). Then there is a long gap of countries that english is likely to be commonly understood, and at the 9th position there is Brasil (2.3%) of users. So having Chinesse translation is definitely a good idea.

File: src/froll.c [hunk 5]
```diff
@@ -45,7 +45,7 @@ void frollmean(unsigned int algo, double *x, uint64_t nx, ans_t *ans, int k, int
  */
 void frollmeanFast(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool narm, int hasna, bool verbose) {
   if (verbose)
-    snprintf(end(ans->message[0]), 500, "frollmeanFast: running for input length %"PRIu64", window %d, hasna %d, narm %d\n", (uint64_t)nx, k, hasna, (int)narm);
```
Comment (MichaelChirico) [inline src/froll.c]: Treat this more as a template so the message doesn't need to be translated redundantly

File: R/data.table.R [hunk 6]
```diff
@@ -820,7 +820,7 @@ replace_dot_alias = function(e) {
           if (!typeof(byval[[jj]]) %chin% ORDERING_TYPES) stop("column or expression ",jj," of 'by' or 'keyby' is type ",typeof(byval[[jj]]),". Do not quote column names. Usage: DT[,sum(colC),by=list(colA,month(colB))]")
         }
         tt = vapply_1i(byval,length)
-        if (any(tt!=xnrow)) stop("The items in the 'by' or 'keyby' list are length (",paste(tt,collapse=","),"). Each must be length ", xnrow, "; the same length as there are rows in x (after subsetting if i is provided).")
+        if (any(tt!=xnrow)) stop(gettextf("The items in the 'by' or 'keyby' list are length(s) (%s). Each must be length %d; the same length as there are rows in x (after subsetting if i is provided).", paste(tt, collapse=","), xnrow, domain='R-data.table'))
```
Comment (MichaelChirico) [inline R/data.table.R]: Converted some messages that are awkward to translate in chunks into one formatted message to be translated as one.

This should probably be done more broadly but it's pretty manual & Chinese matches grammar closely enough. If we ever support Japanese though, probably we'll need complete overhaul...

File: R/setkey.R [hunk 7]
```diff
@@ -251,7 +251,7 @@ setorder = function(x, ..., na.last=FALSE)
 # na.last=FALSE here, to be consistent with data.table's default
 # as opposed to DT[order(.)] where na.last=TRUE, to be consistent with base
 {
-  if (!is.data.frame(x)) stop("x must be a data.frame or data.table.")
+  if (!is.data.frame(x)) stop("x must be a data.frame or data.table")
```
Comment (MichaelChirico) [inline R/setkey.R]: The same message appears elsewhere without `.`, aligning means only one translation needed

**Comment by MichaelChirico** [comment]
Hey @mattdowle we are good to go! All translations completed 🎉

**Comment by mattdowle** [comment]
Under `--as-cran` there's now the following note on the overall `tar.gz` size :
```
* checking CRAN incoming feasibility ... NOTE
Size of tarball: 5129180 bytes
```
v1.12.8 tar.gz size was 4842680 bytes, just under the 5MB CRAN limit.

The `tar.gz` size is not reduced by the compression of the `.Rraw` files already done, because the `tar.gz` is already compressing everything. (The compression of the `.Rraw` files just keeps them compressed when the package is installed to alleviate the "installed size" notes seen previously.)

On next submission we'll see if this new `tar.gz` size note is just on submission, or whether CRAN checks displays it too. We could know that now if there's another package on CRAN with the same "size of tarball" note displayed by its CRAN checks page.  Even if CRAN checks displays that note we'll just have to live with it. I hope the "CRAN ok" badge is still green ok if there are just notes.

**Comment by mattdowle** [comment]
I can't see a NEWS item but maybe I missed it.

File: src/frolladaptive.c [hunk 8]
```diff
@@ -26,13 +26,13 @@ void fadaptiverollmean(unsigned int algo, double *x, uint64_t nx, ans_t *ans, in
  */
 void fadaptiverollmeanFast(double *x, uint64_t nx, ans_t *ans, int *k, double fill, bool narm, int hasna, bool verbose) {
   if (verbose)
-    snprintf(end(ans->message[0]), 500, "fadaptiverollmeanFast: running for input length %"PRIu64", hasna %d, narm %d\n", (uint64_t)nx, hasna, (int) narm);
+    snprintf(end(ans->message[0]), 500, _("%s: running for input length %"PRIu64", hasna %d, narm %d\n"), "fadaptiverollmeanFast", (uint64_t)nx, hasna, (int) narm);
   bool truehasna = hasna>0;                                     // flag to re-run if NAs detected
```
Comment (mattdowle) [inline src/frolladaptive.c]: Ok understood re this change.  Will be interesting if this passes Rtools ATC re PR #4073 and the discussion here: https://github.com/Rdatatable/data.table/pull/4073#discussion_r355211599

Comment (mattdowle) [reply src/frolladaptive.c]: I ran it through win-builder ATC, and yes it passes with no warnings.  So this means it was something about the `__fun__` and not the `%s` before `PRIu64`.  I'll add a link to the previous discussion pointing here too.

Comment (MichaelChirico) [reply src/frolladaptive.c]: awesome, glad to hear

**Comment by MichaelChirico** [comment]
Yep indeed I missed the NEWS item 🤦‍♂ will add a follow-up with tags for all the contributors as well as a summary.

**Pull request merged**